File: 8_String_To_Integer/python3/Solution.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    Solution.py                                        :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: inazaria <inazaria@student.42.fr>          +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/06/16 21:29:02 by inazaria          #+#    #+#              #
#    Updated: 2024/06/16 21:54:49 by inazaria         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

import sys

class Solution:
    def myAtoi(self, s: str) -> int:
        sign: int = 1
        res : int = 0
        idx : int = 0 
        len_s: int = len(s)

        while idx < len_s and s[idx] == " " :
            idx += 1
                
        if 0 <= idx < len_s and s[idx] == "+" :
            idx += 1

        elif 0 <= idx < len_s and s[idx] == "-" :
            sign *= -1
            idx += 1        

        while idx < len(s) and ord(s[idx]) >= 48 and ord(s[idx]) <= 57 :
            res = 10 * res + ord(s[idx]) - 48
            idx += 1 

        res = res * sign
        res =  2 ** 31 - 1 if res >  2 ** 31 - 1 else res
        res = -2 ** 31     if res < -2 ** 31 else res

        return res
        


if __name__ == "__main__" : 
    if (len(sys.argv) != 2) :
        print("Usage: python3 Solution.py <string>")
        print("Example: python3 Solution.py \"42\"")
        sys.exit(1)

    try :
        num = str(sys.argv[1])
 
    except Exception as e:
        print("Error: ", e)
        print("Usage: python3 Solution.py <string>")
        print("Example: python3 Solution.py \"42\"")
        sys.exit(1)
    
    answer = Solution().myAtoi(num)
    print(f"atoi({num}) = {answer}")

File: 8_String_To_Integer/C/Solution.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   Solution.c                                         :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: inazaria <inazaria@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/06/16 21:55:01 by inazaria          #+#    #+#             */
/*   Updated: 2024/06/16 22:35:25 by inazaria         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

# include <stdio.h>
# include <limits.h>

int myAtoi(char* s) {
	int		sign = 1;
	int		idx = 0;
	long	res = 0;

	while (s[idx] && s[idx] == ' ') {
		idx++;
	}

	if (s[idx] && s[idx] == '+') {
		idx++;
	}

	else if (s[idx] && s[idx] == '-') {
		sign = -sign;
		idx++;
	}

	while (s[idx] && s[idx] >= '0' && s[idx] <= '9') {
		
		res = 10 * res + s[idx] - '0';
		if (res * sign > INT_MAX) {
			res = INT_MAX;
			return res;
		}
		if (res * sign < INT_MIN) {
			res = INT_MIN;
			return res;
		}	
		idx++;
	}

	res *= sign;
	return (res);	
}

int	main(int argc, char *argv[]) {
	if (argc != 2) {
		fprintf(stdout, "Usage : ./a.out \"num\"");
		return (1);
	}
	int	answer;
	answer = myAtoi(argv[1]);
	printf("myAtoi(%s) = %d", argv[1], answer);
	return (0);
}
